docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,208 @@
+chain processing.
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+
+func newRecoveryMiddleware(handler RecoveryHandler, next recoveryMiddleware) recoveryMiddleware {
+    return func(recoveryObj interface{}) (recoveryMiddleware, error) {
+        if err := handler(recoveryObj); err != nil {
+            return nil, err
+        }
+        return next, nil
+    }
+}
+```
+
+Function receives a `recoveryObj` object and returns:
+* (next `recoveryMiddleware`, `nil`) if object wasn't handled (not a target type) by `RecoveryHandler`;
+* (`nil`, not nil `error`) if input object was handled and other middlewares in the chain should not be executed;
+* (`nil`, `nil`) this is an invalid behaviour 'cause in that case panic recovery might not be properly handled;
+this can be avoided by always using a `default` as a rightmost middleware in chain (always returns an `error`'); 
+
+
+`OutOfGas` middleware example:
+```go
+func newOutOfGasRecoveryMiddleware(gasWanted uint64, ctx sdk.Context, next recoveryMiddleware) recoveryMiddleware {
+    handler := func(recoveryObj interface{}) error {
+        err, ok := recoveryObj.(sdk.ErrorOutOfGas)
+        if !ok { return nil }
+
+        return sdkerrors.Wrap(
+            sdkerrors.ErrOutOfGas, fmt.Sprintf(
+                "out of gas in location: %v; gasWanted: %d, gasUsed: %d", err.Descriptor, gasWanted, ctx.GasMeter().GasConsumed(),
+            ),
+        )
+    }
+    
+    return newRecoveryMiddleware(handler, next)
+}
+```
+
+`Default` middleware example:
+```go
+func newDefaultRecoveryMiddleware() recoveryMiddleware {
+    handler := func(recoveryObj interface{}) error {
+        return sdkerrors.Wrap(
+            sdkerrors.ErrPanic, fmt.Sprintf("recovered: %v\nstack:\n%v", recoveryObj, string(debug.Stack())),
+        )
+    }
+    
+    return newRecoveryMiddleware(handler, nil)
+}
+```
+
+##### Recovery processing
+
+Basic chain of middlewares processing would look like:
+
+```go
+func processRecovery(recoveryObj interface{}, middleware recoveryMiddleware) error {
+	if middleware == nil { return nil }
+
+	next, err := middleware(recoveryObj)
+	if err != nil { return err }
+	if next == nil { return nil }
+
+	return processRecovery(recoveryObj, next)
+}
+```
+
+That way we can create a middleware chain which is executed from left to right, the rightmost middleware is a
+`default` handler which must return an `error`.
+
+##### BaseApp changes
+
+The `default` middleware chain must exist in a `BaseApp` object. `Baseapp` modifications:
+
+```go
+type BaseApp struct {
+    // ...
+    runTxRecoveryMiddleware recoveryMiddleware
+}
+
+func NewBaseApp(...) {
+    // ...
+    app.runTxRecoveryMiddleware = newDefaultRecoveryMiddleware()
+}
+
+func (app *BaseApp) runTx(...) {
+    // ...
+    defer func() {
+        if r := recover(); r != nil {
+            recoveryMW := newOutOfGasRecoveryMiddleware(gasWanted, ctx, app.runTxRecoveryMiddleware)
```

Comment:
Why are we hard-coding `newOutOfGasRecoveryMiddleware` into `BaseApp`? I thought the idea was to defer to `app.runTxRecoveryMiddleware`?

Reply:
That is because `OutOfGas` error handling needs some extra context and can't be statically created.